improver/bases.py

```diff
@@ -0,0 +1,261 @@
+        """
+        cube = getattr(self, f"_{key}")
+        if not isinstance(cube, Cube):
+            raise TypeError(f"_{key} should be a Cube, but found {type(cube)}")
+        return cube
+
+    @staticmethod
+    def assert_time_coords_ok(inputs: List[Cube], time_bounds: bool):
+        """
+        Raises appropriate ValueError if
+
+        - Any input cube has or is missing time bounds (depending on time_bounds)
+        - Input cube times and forecast_reference_times do not match
+
+        Can be overloaded where only a subset of inputs are expected to match, or have
+        specific offsets. Overloading functions can use self.get_cube()
+        """
+        if len(inputs) <= 1:
+            return
+        cubes_not_matching_time_bounds = [
+            c.name() for c in inputs if c.coord("time").has_bounds() != time_bounds
+        ]
+        if cubes_not_matching_time_bounds:
+            str_bool = "" if time_bounds else "not "
+            msg = f"{' and '.join(cubes_not_matching_time_bounds)} must {str_bool}have time bounds"
+            raise ValueError(msg)
+        for time_coord_name in ["time", "forecast_reference_time"]:
+            time_coords = [c.coord(time_coord_name) for c in inputs]
+            if not all([tc == time_coords[0] for tc in time_coords[1:]]):
+                raise ValueError(
+                    f"{time_coord_name} coordinates do not match."
+                    "\n  "
+                    "\n  ".join(
+                        [str(c.name()) + ": " + str(c.coord("time")) for c in inputs]
+                    )
+                )
+
+    @staticmethod
+    def assert_spatial_coords_ok(spatial_matching_cubes: List[Cube]):
+        """
+        Raises appropriate ValueError if
+
+        - if the x and y coords are not exactly the same to the
+            precision of the floating-point values (this should be true for
+            any cubes derived using cube.regrid())
+
+        Can be overloaded where only a subset of inputs are expected to match.
+        Overloading functions can use self.get_cube()
+        """
+        if len(spatial_matching_cubes) <= 1:
+            return
+        if not spatial_coords_match(spatial_matching_cubes):
+            raise ValueError(
+                f"Spatial coords of input Cubes do not match: {spatial_matching_cubes}"
+            )
+
+    def parse_inputs(self, inputs: List[Cube], time_bounds=False) -> None:
+        """Extracts input cubes as described by self.cube_descriptors.
+
+        Args:
+            inputs:
+                List of Cubes containing exactly one of each input cube.
+            time_bounds:
+                True when all input cubes are expected to hate time bounds.
+        Raises:
+            ValueError:
+                If additional cubes are found
+        """
+        cubes = CubeList(inputs)
+        for desc in self.cube_descriptors.values():
+            desc._matched_name = desc.name
```

Comment:
I think the purpose of `cube_descriptors` is maybe a little convoluted here as are writing data to it (matched_name).
`cube_descrptors` is intended as a container for our plugins cube constraints.
This suggests to me the current workflow might not be quite right.